extra/DEVELOPER_DOCS/StringStore-Vocab.md

```diff
@@ -0,0 +1,212 @@
+
+In Python 2 strings were more like sequences of bytes, but in Python 3 they're
+more like UTF-8 sequences and treated separately from bytes. Cython has several
+string types to manage compatability with bytes and strings in Python 2 and 3:
+`str`, `unicode`, `bytes`, and `basestring`. 
+
+Here is how the types work in Python 3:
+
+- `str` and `unicode` are the same
+- `bytes` is just `bytes`
+- `basestring` is the union of `str` and `unicode`, so in Py3 all are equivalent
+
+`basestring` and `unicode` are still used in several places in spaCy code,
+presumably as a legacy of Python 2 support. In Python 3 there should be no
+difference between these types and `str`.
+
+### When to Use the StringStore?
+
+While you can ignore the `StringStore` in many cases, there are situations where
+you should make use of it to avoid errors. 
+
+Any time you introduce a string that may be set on a Doc field that has a hash,
+you should add the string to the `StringStore`. This mainly happens when adding
+labels in components, but there are some other cases:
+
+- syntax iterators, mainly `get_noun_chunks`
+- external data used in components, like the KnowledgeBase in EntityLinking
+- labels used in tests
+
+## Vocab
+
+The `Vocab` is a core component of a `Language` pipeline. Its main function is
+to manage `Lexeme`s, which are structs that contain information about a token
+that depends only on its surface form, without context. `Lexeme`s store much of
+the data associated with `Token`s. As a side effect of this the `Vocab` also
+managed the `StringStore` for a pipeline and a grab-bag of other data.
+
+These are things stored in the vocab:
+
+- `Lexeme`s
+- `StringStore`
+- `Morphology`: manages info used in `MorphAnalysis` objects
+- `vectors`: basically a dict for word vectors
+- `lookups`: language specific data like lemmas
+- `writing_system`: language specific metadata
+- `get_noun_chunks`: a syntax iterator
+- `data_dir`: unused?
+- lex attribute getters: functions like `is_punct`, set in language defaults
+- `cfg`: **not** the pipeline config, this is mostly unused
+
+Some of these, like the Morphology and Vectors, are complex enough that they
+need their own explanations. Here we'll just look at Vocab-specific items.
+
+### Lexemes
+
+A `Lexeme` is a type that mainly wraps a `LexemeC`, a struct consisting of ints
+that identify various context-free token attributes. Lexemes are the core data
+of the `Vocab`, and can be accessed using `__getitem__` on the `Vocab`. The memory
+for storing `LexemeC` objects is managed by a pool that belongs to the `Vocab`.
+
+The attributes stored in a `LexemeC` are:
+
+- orth (the raw text)
+- lower
+- norm
+- shape
+- prefix
+- suffix
+
+Most of these are straightforward. `shape` can technically be customized,
+though that's not usually done. `norm` is more likely to be customized, and
+spaCy doesn't prescribe any particular type of normalization.
+
```

Comment:
These can all be customized in the language defaults (even `lower`). In particular, the prefix and suffix lengths are probably good for English and not optimal for many other languages, but currently there's no much language-specific customization.